### DISCUSSION TITLE
HL7800 Modem FW Update(UART TRANSFER SPEED) 3 TIMES SLOWER THAN TERA TERM TRANSFER

### DISCUSSION THREAD

**Discussion by zafersn** [zephyrproject-rtos/zephyr#81379] [Q&A]
Hi All,

I recently performed an experiment with a HL7800 modem firmware update using the delta MDM firmware package. I transferred the hardcoded delta firmware package to the modem over UART using the HL7800 driver at a baud rate of 115200 with the modem X 1K protocol. The transfer of 530,830 bytes took approximately 43 minutes (transfer speed: 204 bytes/sec) with the HL7800 driver. However, when I transferred the same file over UART (using a TTL converter to access the UART port at the same baud rate) using Tera Term, it took around 15 minutes (transfer speed: 640 bytes/sec).

1. **Has anyone else encountered this issue or have any suggestions? Here is a comparison of the transfer speeds between Tera Term and the hardcoded HL7800 modem driver.**
2. **Do you think using DMA for UART communication would have a positive effect?**

HL7800 driver:
![image](https://github.com/user-attachments/assets/41975d4e-5455-44bc-9d18-b2484e14f763)

TERA term:
![image](https://github.com/user-attachments/assets/e5ab34ab-445c-4611-9a09-e473a8337163)

**Comment by Sambsc19** [comment]
Hey buddy, is it PC->UART->MCU->UART->MODEM?

We need to identify where the bottleneck is. Is it PC->MCU or MCU->MODEM. The binary can be written pc->stm32 flash mem when programming FW without any noticeable delay. Just came across this by chance. Ill share cmd monday if you need it. 43min is insane

**Reply by zafersn** [reply]
> Just came across this by chance

Haha, that means today is my lucky day. Thanks for your reply, @Sambsc19 

> We need to identify where the bottleneck is. Is it PC->MCU or MCU->MODEM. 

I ran two different experiments:

1 .MCU->MODEM takes ~43 mins
I had a delta modem upgrade .ua file, so for experimental purposes, I converted it to hex and hardcoded it into my project to test the transfer speed. I used the exact firmware update structure of the HL7800 driver to transmit the firmware to the modem, except for the filesystem part (I included the firmware as a header file, so I didn't need to use the filesystem).

Here is what the converted .ua file looks like:
![image](https://github.com/user-attachments/assets/c44f173f-4d81-4b80-88a0-896c72ba94c4)

Did you also try to update your modem via the MCU? If so, what were the transmit speed and time?

2. PC -> Modem takes ~15mins
In this setup, I directly connected the modem's UART1 TX and RX pins to the PC using a TTL converter and then ran Tera Term.

> The binary can be written pc->stm32 flash mem when programming FW without any noticeable delay. Just came across this by chance. Ill share cmd monday if you need it. 43min is insane

To transfer the .ua file, I first converted it to a header file using a bash command

```
#!/bin/bash
# Convert the firmware file to a plain hex dump
xxd -p HL7812_5.5.8.0_LE_23.10.0_to_5.5.14.0_LE.ua | tr -d '\n' | sed 's/\(..\)/0x\1, /g' | fold -s -w 80 | sed '1s/^/#ifndef FIRMWARE_UPDATE_IMAGE_H\n#define FIRMWARE_UPDATE_IMAGE_H\n\nconst unsigned char firmware_update_image[] = {\n/' | sed '$s/, $/\n};\n\n#endif \/\/ FIRMWARE_UPDATE_IMAGE_H\n/' > firmware_update_image_8_to_14.h

```

Then, I included the generated header file in the project. That’s it. As you mentioned, flashing the hex file with the hardcoded modem firmware from the PC to the STM32 takes just a few seconds. However, transferring it from the MCU to the modem is extremely slow.

> Ill share cmd monday if you need it

whats that cmd? i didnt get it. but yeah sure please share it.
